lambda/audit.py

```diff
@@ -0,0 +1,319 @@
+import json
+import os
+import uuid
+
+import boto3
+from botocore.exceptions import ClientError
+
+import github_api
+from logger import LOG
+
+
+def start(message=None):
+    """ Start an audit of github membership """
+    org = os.environ.get("GITHUB_ORG")
+    audit_id = str(uuid.uuid4())
+    LOG.info({"action": "Starting github audit", "org": org, "audit_id": audit_id})
+
+    sns_org_members = create_sns_message(
+        {"action": "log_org_membership", "org": org, "audit_id": audit_id}
+    )
+    publish_alert(sns_org_members)
+
+    sns_org_repos = create_sns_message(
+        {"action": "log_org_repos", "org": org, "audit_id": audit_id}
+    )
+    publish_alert(sns_org_repos)
+
+    sns_org_teams = create_sns_message(
+        {"action": "log_org_teams", "org": org, "audit_id": audit_id}
+    )
+    publish_alert(sns_org_teams)
+    return True
+
+
+def log_org_membership(message=None):
+    """ Audit github organization membership """
+    org = os.environ.get("GITHUB_ORG")
+    audit_id = message.get("audit_id")
+    if audit_id is not None:
+        LOG.info({"action": "Audit organization members", "org": org})
+        members = github_api.get_github_org_members(org)
+        event = {
+            "type": "OrganizationMemberCount",
+            "org": org,
+            "count": len(members),
+            "audit_id": audit_id,
+        }
+        LOG.info(event)
+        for member in members:
+            event = {
+                "type": "OrganizationMember",
+                "org": org,
+                "member": member,
+                "team": None,
+                "repository": None,
+                "audit_id": audit_id,
+            }
+            LOG.info(event)
+    return True
+
+
+def log_org_teams(message=None):
+    """ Audit github organization teams """
+    org = os.environ.get("GITHUB_ORG")
+    audit_id = message.get("audit_id")
+    if audit_id is not None:
+        LOG.info({"action": "Audit organization teams", "org": org})
+        teams = github_api.get_github_org_teams(org)
+        event = {
```

Comment:
Should these event creations be factored out into their own function? Is there a standard event format? Why do some events have some fields but others don't. What does `None` mean in the context of an event?

Reply:
Ian wanted the data flat to make it easier to write the splunk searches. 
The idea is that for example 
repo + member + team = access via team (you need to remove a leaver from the team) 
repo + member + NO team = direct contributor (you need to remove a leaver from the repo settings)

Would probably make sense to wrap it in a function with keyword args I guess.

Reply:
Thanks for splitting these out.